File: Samples/Mapping/Maps-Samples/POI/POIDetailsViewController.m
//
//  POIDetailsViewController.m
//  PWMapKit
//
//  Created on 8/10/16.
//  Copyright © 2016 Phunware. All rights reserved.
//

#import <PWMapKit/PWMapKit.h>
#import <PureLayout/PureLayout.h>
#import <QuartzCore/QuartzCore.h>

#import "POIDetailsViewController.h"
#import "CommonSettings.h"
#import "PreRoutingViewController.h"
#import "POIDetailsTableViewCell.h"
#import "POIDetailsSummaryCell.h"

static CGFloat const GetDirectionsContainerViewHeight = 80.0;

@interface POIDetailsViewController () <UITableViewDelegate, UITableViewDataSource>

@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) UIView *getDirectionsContainerView;

@end

@implementation POIDetailsViewController

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.navigationItem setTitle:self.pointOfInterest.title];
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    self.navigationItem.title = PWLocalizedString(@"BackButtonTitle", @"Back");
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self initializeAndAddAllSubviews];
    [self configureTableView];
    [self configureGetDirectionsContainerView];
}

#pragma mark - Configuration

- (void)initializeAndAddAllSubviews {
    self.tableView = [UITableView new];
    self.getDirectionsContainerView = [UIView new];
    
    [self.view addSubview:self.tableView];
    [self.view addSubview:self.getDirectionsContainerView];
}

- (void)configureTableView {
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    self.tableView.contentInset = UIEdgeInsetsMake(0, 0, GetDirectionsContainerViewHeight, 0);
    self.tableView.scrollIndicatorInsets = UIEdgeInsetsMake(0, 0, GetDirectionsContainerViewHeight, 0);
    
    [self.tableView registerClass:[POIDetailsTableViewCell class] forCellReuseIdentifier:POIDetailsCellReuseIdentifier];
    [self.tableView registerClass:[POIDetailsSummaryCell class] forCellReuseIdentifier:POIDetailsSummaryCellReuseIdentifier];
    
    [self.tableView autoPinEdgesToSuperviewEdges];
}

- (void)configureGetDirectionsContainerView {
    self.getDirectionsContainerView.backgroundColor = [UIColor clearColor];
    
    [self.getDirectionsContainerView autoPinEdgeToSuperviewEdge:ALEdgeBottom];
    [self.getDirectionsContainerView autoPinEdgeToSuperviewEdge:ALEdgeLeft];
    [self.getDirectionsContainerView autoPinEdgeToSuperviewEdge:ALEdgeRight];
    if (!self.userLocation) {
        self.getDirectionsContainerView.hidden = YES;
        [self.getDirectionsContainerView autoSetDimension:ALDimensionHeight toSize:0];
    } else {
        self.getDirectionsContainerView.hidden = NO;
        [self.getDirectionsContainerView autoSetDimension:ALDimensionHeight toSize:GetDirectionsContainerViewHeight];
    }
    
    UIButton *getDirectionsButton = [UIButton new];
    getDirectionsButton.backgroundColor = [CommonSettings commonButtonBackgroundColor];
    getDirectionsButton.tintColor = [UIColor whiteColor];
    getDirectionsButton.layer.cornerRadius = 25;
    getDirectionsButton.clipsToBounds = YES;
    [getDirectionsButton setTitle:PWLocalizedString(@"GetDirectionsButtonTitle", @"GET DIRECTIONS") forState:UIControlStateNormal];
    getDirectionsButton.titleLabel.font = [UIFont boldSystemFontOfSize:20.0];
    [getDirectionsButton addTarget:self action:@selector(getDirectionsButtonTapped) forControlEvents:UIControlEventTouchUpInside];
    [self.getDirectionsContainerView addSubview:getDirectionsButton];
    
    [getDirectionsButton autoAlignAxisToSuperviewAxis:ALAxisHorizontal];
    [getDirectionsButton autoAlignAxisToSuperviewAxis:ALAxisVertical];
    [getDirectionsButton autoMatchDimension:ALDimensionHeight toDimension:ALDimensionHeight ofView:self.getDirectionsContainerView withMultiplier:0.6];
    [getDirectionsButton autoMatchDimension:ALDimensionWidth toDimension:ALDimensionWidth ofView:self.getDirectionsContainerView withMultiplier:0.9];
}

#pragma mark - Table View DataSource and Delegate

- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 44;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return UITableViewAutomaticDimension;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 5;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = nil;
    
    switch (indexPath.row) {
        case 0: {
            POIDetailsTableViewCell *detailsCell = [tableView dequeueReusableCellWithIdentifier:POIDetailsCellReuseIdentifier forIndexPath:indexPath];
            [detailsCell configureForKey:PWLocalizedString(@"BuildingColon", @"Building:") value:self.pointOfInterest.floor.building.name];
            cell = detailsCell;
        }
            break;
        case 1: {
            POIDetailsTableViewCell *detailsCell = [tableView dequeueReusableCellWithIdentifier:POIDetailsCellReuseIdentifier forIndexPath:indexPath];
            [detailsCell configureForKey:PWLocalizedString(@"FloorColon", @"Floor:") value:self.pointOfInterest.floor.name];
            cell = detailsCell;
        }
            break;
        case 2: {
            POIDetailsTableViewCell *detailsCell = [tableView dequeueReusableCellWithIdentifier:POIDetailsCellReuseIdentifier forIndexPath:indexPath];
            [detailsCell configureForKey:PWLocalizedString(@"TypeColon", @"Type:") value:self.pointOfInterest.pointOfInterestType.name];
            cell = detailsCell;
        }
            break;
        case 3: {
            POIDetailsTableViewCell *detailsCell = [tableView dequeueReusableCellWithIdentifier:POIDetailsCellReuseIdentifier forIndexPath:indexPath];
            [detailsCell configureForKey:PWLocalizedString(@"CoordinateColon", @"Coordinate:") value:[NSString stringWithFormat:@"%.6f,%.6f", self.pointOfInterest.coordinate.latitude, self.pointOfInterest.coordinate.longitude]];
            cell = detailsCell;
        }
            break;
        case 4: {
            POIDetailsSummaryCell *summaryCell = [tableView dequeueReusableCellWithIdentifier:POIDetailsSummaryCellReuseIdentifier forIndexPath:indexPath];
            [summaryCell configureForSummary:self.pointOfInterest.summary];
            cell = summaryCell;
        }
            break;
        default: {
            POIDetailsTableViewCell *detailsCell = [tableView dequeueReusableCellWithIdentifier:POIDetailsCellReuseIdentifier forIndexPath:indexPath];
            [detailsCell configureForKey:@"" value:@""];
            cell = detailsCell;
        }
            break;
    }
    
    return cell;
}

#pragma mark - UI Actions

- (void)getDirectionsButtonTapped {
    __weak typeof(self) weakSelf = self;
    [PWRoute initRouteFrom:self.userLocation to:self.pointOfInterest accessibility:YES completion:^(PWRoute *route, NSError *error) {
        PreRoutingViewController *preRoutingViewController = [PreRoutingViewController new];
        preRoutingViewController.route = route;
        [weakSelf showViewController:preRoutingViewController sender:weakSelf];
    }];
}

- (void)backButtonTapped {
    [self.navigationController popToRootViewControllerAnimated:YES];
}

@end

File: Samples/PhunMaps/PhunMaps/Controllers/POIDetailsViewController.swift
//
//  POIDetailsViewController.swift
//  Mapping-Sample
//
//  Created on 6/13/17.
//  Copyright © 2017 Phunware. All rights reserved.
//

import UIKit
import PWMapKit

class POIDetailsViewController: UIViewController {
    
    @IBOutlet weak var tableView: UITableView!    
    @IBOutlet weak var directionsButton: UIButton!
    
    var pointOfInterest: PWPointOfInterest?
    var userLocation: PWUserLocation?
    var route: PWRoute?
    
    fileprivate var rowMapping = [rowMap]()
    
    enum rowMap: Int {
        case building
        case floor
        case type
        case coordinate
        case summary
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.rowHeight = UITableViewAutomaticDimension
        tableView.estimatedRowHeight = 55
        
        configureRowMapping()
        
        directionsButton.layer.cornerRadius = 25
        directionsButton.clipsToBounds = true
        directionsButton.backgroundColor = CommonSettings.buttonBackgroundColor
        directionsButton.setTitle(NSLocalizedString("GET DIRECTIONS", comment: ""), for: .normal)
        directionsButton.addTarget(self, action: #selector(getDirectionsPressed(sender:)), for: .touchUpInside)
        
        directionsButton.isHidden = userLocation == nil
    }
    
    func configureRowMapping() {
        rowMapping.append(.building)
        rowMapping.append(.floor)
        rowMapping.append(.type)
        rowMapping.append(.coordinate)
        rowMapping.append(.summary)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        navigationController?.navigationBar.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.white]
        if let pointOfInterest = pointOfInterest {
            title = pointOfInterest.title
        }
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
    }
    
    @objc func getDirectionsPressed(sender: UIButton) {
        guard let userLocation = userLocation, let pointOfInterest = pointOfInterest else {
            return
        }
        
        PWRoute.createRoute(from: userLocation, to: pointOfInterest, accessibility: false, excludedPoints: nil) { [weak self] (route, error) in
            if let error = error {
                print(error.localizedDescription)
                return
            }
            
            if route == nil {
                let alertController = UIAlertController(title: NSLocalizedString("You have arrived!", comment: ""), message: NSLocalizedString("No route found because your location is too close to the destination", comment: ""), preferredStyle: .alert)
                let okayAction = UIAlertAction(title: "Okay", style: .default, handler: { [weak self] (action) in
                    self?.dismiss(animated: true, completion: nil)
                })
                
                alertController.addAction(okayAction)
                
                self?.present(alertController, animated: true, completion: nil)
            }
            
            self?.route = route
            self?.performSegue(withIdentifier: String(describing: PreRoutingViewController.self), sender: nil)
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard let identifier = segue.identifier else {
            return
        }
        
        if identifier == String(describing: PreRoutingViewController.self), let destination = segue.destination as? PreRoutingViewController {
            destination.route = route
        }
    }
}

extension POIDetailsViewController: UITableViewDelegate { }

extension POIDetailsViewController: UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return rowMapping.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell = UITableViewCell()
        
        switch rowMapping[indexPath.row] {
        case .building:
            if let detailsCell = tableView.dequeueReusableCell(withIdentifier: String(describing: POIDetailsCell.self), for: indexPath) as? POIDetailsCell {
                detailsCell.titleLabel.text = NSLocalizedString("Building:", comment: "")
                
                if let buildingName = pointOfInterest?.floor.building.name {
                    detailsCell.detailLabel.text = buildingName
                }
                
                cell = detailsCell
            }
        case .floor:
            if let detailsCell = tableView.dequeueReusableCell(withIdentifier: String(describing: POIDetailsCell.self), for: indexPath) as? POIDetailsCell {
                detailsCell.titleLabel.text = NSLocalizedString("Floor:", comment: "")
                if let floorName = pointOfInterest?.floor.name {
                    detailsCell.detailLabel.text = floorName
                }
                
                cell = detailsCell
            }
        case .type:
            if let detailsCell = tableView.dequeueReusableCell(withIdentifier: String(describing: POIDetailsCell.self), for: indexPath) as? POIDetailsCell {
                detailsCell.titleLabel.text = NSLocalizedString("Type:", comment: "")
                if let poiTypeName = pointOfInterest?.pointOfInterestType.name {
                    detailsCell.detailLabel.text = poiTypeName
                }
                
                cell = detailsCell
            }
        case .coordinate:
            if let detailsCell = tableView.dequeueReusableCell(withIdentifier: String(describing: POIDetailsCell.self), for: indexPath) as? POIDetailsCell {
                detailsCell.titleLabel.text = NSLocalizedString("Coordinate:", comment: "")
                if let pointOfInterest = pointOfInterest {
                    detailsCell.detailLabel.text = "\(pointOfInterest.coordinate.latitude), \(pointOfInterest.coordinate.longitude)"
                }
                
                cell = detailsCell
            }
        case .summary:
            if let detailsSummaryCell = tableView.dequeueReusableCell(withIdentifier: String(describing: POIDetailsSummaryCell.self), for: indexPath) as? POIDetailsSummaryCell {
                if let summary = pointOfInterest?.summary {
                    detailsSummaryCell.summaryLabel.text = summary
                }
                
                cell = detailsSummaryCell
            }
        }
        
        return cell
    }
}
